pkg/mempool/custom_proposal_handler_test.go

```diff
@@ -0,0 +1,393 @@
+		{tx: buildEthMsg(s.T(), txConfig, []byte(`13`), [][]byte{secret1}, []uint64{5}), priority: 10},
+		{tx: buildEthMsg(s.T(), txConfig, []byte(`14`), [][]byte{secret1}, []uint64{6}), priority: 8},
+		// large cosmos tx to test with eth txs
+		{
+			tx: buildMsg(
+				s.T(),
+				txConfig,
+				[]byte(`1252345678910125234567891012523456789101252345678910125234567891012523456789101252345678910125234567891012523456789101252345678912343`),
+				[][]byte{secret2},
+				[]uint64{1},
+			),
+			priority: 8,
+		},
+	}
+
+	for i := range testTxs {
+		bz, err := txConfig.TxEncoder()(testTxs[i].tx)
+		s.Require().NoError(err)
+		testTxs[i].bz = bz
+		testTxs[i].size = int(cmttypes.ComputeProtoSizeForTxs([]cmttypes.Tx{bz}))
+	}
+
+	s.Require().Equal(testTxs[0].size, 111)
+	s.Require().Equal(testTxs[1].size, 121)
+	s.Require().Equal(testTxs[2].size, 112)
+	s.Require().Equal(testTxs[3].size, 112)
+	s.Require().Equal(testTxs[4].size, 195)
+	s.Require().Equal(testTxs[5].size, 205)
+	s.Require().Equal(testTxs[6].size, 196)
+	s.Require().Equal(testTxs[7].size, 196)
+	s.Require().Equal(testTxs[8].size, 196)
+
+	// ethermint txs
+	s.Require().Equal(testTxs[15].size, 247)
+	s.Require().Equal(testTxs[16].size, 257)
+	s.Require().Equal(testTxs[17].size, 248)
+	s.Require().Equal(testTxs[18].size, 248)
+
+	s.Require().Equal(testTxs[19].size, 259)
+	s.Require().Equal(testTxs[20].size, 248)
+	s.Require().Equal(testTxs[21].size, 248)
+
+	s.Require().Equal(testTxs[22].size, 248)
+
+	testCases := map[string]struct {
+		ctx         sdk.Context
+		txInputs    []testTx
+		req         abci.RequestPrepareProposal
+		handler     sdk.PrepareProposalHandler
+		expectedTxs []int
+	}{
+		"skip same-sender non-sequential sequence and then add others txs": {
+			ctx:      s.ctx,
+			txInputs: []testTx{testTxs[0], testTxs[1], testTxs[2], testTxs[3]},
+			req: abci.RequestPrepareProposal{
+				MaxTxBytes: 111 + 112,
+			},
+			expectedTxs: []int{0, 3},
+		},
+		"(eth tx) skip same-sender non-sequential sequence and then add others txs": {
+			ctx:      s.ctx,
+			txInputs: []testTx{testTxs[15], testTxs[16], testTxs[17], testTxs[18]},
+			req: abci.RequestPrepareProposal{
+				MaxTxBytes: 247 + 248,
+			},
+			expectedTxs: []int{15, 18},
+		},
+		"(eth tx + cosmos tx) skip same-sender non-sequential sequence and then add others txs": {
+			ctx:      s.ctx,
+			txInputs: []testTx{testTxs[15], testTxs[16], testTxs[17], testTxs[22]},
+			req: abci.RequestPrepareProposal{
+				MaxTxBytes: 247 + 248,
+			},
+			expectedTxs: []int{15, 22},
+		},
+		"skip multi-signers msg non-sequential sequence": {
+			ctx:      s.ctx,
+			txInputs: []testTx{testTxs[4], testTxs[5], testTxs[6], testTxs[7], testTxs[8]},
+			req: abci.RequestPrepareProposal{
+				MaxTxBytes: 195 + 196,
+			},
+			expectedTxs: []int{4, 8},
+		},
+		"only the first tx is added": {
+			// Because tx 10 is valid, tx 11 can't be valid as they have higher sequence numbers.
```

Comment:
Using index in the comments as well might be better .

Reply:
yes but also the comment in that case should be also be `Because tx 10 is invalid` ?